File: proj_windowccl/py/wccl.py
"""
Window Connected Component Labeling (WCCL)

The python code here is not meant to be efficient within Python, but rather to mirror the logic
that should be implemented inside the eventual CUDA code. Specific points to note:

- The connections for the index map are made on a copy of the input. This is because in any CUDA implementation,
it will be impossible to ensure the order of the the parent traversal for active sites. As such, there is no way
to ensure that the connections from another element can be seen by the current element. Hence, here in the python
implementation we choose to *ignore* all prior elements' writes by reading from the input but writing to a copy.
"""

import numpy as np

def parse_string_to_binary_map(s: list[str]):
    """
    Generates a binary map from a list of equal length strings

    Parameters
    ----------
    s : list[str]
        List of equal length strings.
        E.g. [
            "010001",
            "100010",
        ]

    Returns
    -------
    binary_map : np.ndarray
        np.uint8 binary map of equal shape
    """
    # Check that each string is equal length
    if not all(len(i) == len(s[0]) for i in s):
        raise ValueError("All strings must be equal length")

    binary_map = np.array([
        [int(i) for i in row]
        for row in s
    ], dtype=np.uint8)
    return binary_map

def rand_binary_map(rows: int, cols: int) -> np.ndarray:
    """
    Generates a random binary map, usually for testing.

    Parameters
    ----------
    rows : int
        Number of rows

    cols : int
        Number of columns

    Returns
    -------
    binary_map : np.ndarray
        np.uint8 binary map of shape (rows, cols)
    """
    return np.random.randint(low=0, high=2, size=(rows, cols)).astype(np.uint8)

def make_indices_map(binary_map: np.ndarray, dtype: type = np.int32) -> np.ndarray:
    """
    Converts the binary map to an index map, used as a starting point for connections;
    Active sites (1) receive their flattened 1D index, while inactive sites (0) receive -1.

    Parameters
    ----------
    binary_map : np.ndarray
        Input binary map.

    dtype : type
        Data type of the index map. Defaults to np.int32.

    Returns
    -------
    idx : np.ndarray
        Index map.
    """
    if binary_map.dtype != np.uint8:
        raise ValueError("binary_map must be uint8")
    if np.any(binary_map > 1):
        raise ValueError("binary_map must contain only 1s or 0s")
    idx = np.arange(binary_map.size).reshape(binary_map.shape).astype(dtype)
    idx[binary_map == 0] = -1
    return idx


def make_connections(idx_map: np.ndarray, hdist: int, vdist: int):
    numChanges = 0
    workspace = idx_map.copy()
    for i in range(0, idx_map.shape[0]):
        for j in range(0, idx_map.shape[1]):
            if (idx_map[i,j] < 0):
                continue
            # Extract window around current point
            minrow = max(0, i - hdist)
            maxrow = min(idx_map.shape[0], i + hdist + 1) # non-inclusive
            mincol = max(0, j - vdist)
            maxcol = min(idx_map.shape[1], j + vdist + 1) # non-inclusive
            # print(f"Index ({i}, {j}) has window [{minrow}, {maxrow}, {mincol}, {maxcol}]")
            window_idx = idx_map[minrow:maxrow, mincol:maxcol]
            # print(window_idx)
            root = np.min(window_idx[window_idx >= 0])

            if (workspace[i,j] != root):
                print(f"Setting idx[{i}, {j}] = {root}")
                workspace[i,j] = root
                numChanges += 1

    return workspace, numChanges

def connect(idx_map: np.ndarray, hdist: int, vdist: int, maxIter: int = 100):
    numChanges_list = []
    numIter = 0
    while numIter < maxIter:
        idx_map, numChanges = make_connections(idx_map, hdist, vdist)
        numChanges_list.append(numChanges)
        if numChanges == 0:
            break

        print(idx_map)
        numIter += 1

    if numChanges_list[-1] != 0:
        raise ValueError(f"Failed to connect indices within {maxIter} iterations")

    return idx_map, numChanges_list

File: proj_windowccl/src/wccl.cpp
#include "wccl.h"

#include <stdexcept>
#include <algorithm>

namespace wccl
{

CPUSolver::CPUSolver(const uint32_t hDist, const uint32_t vDist)
  : m_hDist(hDist), m_vDist(vDist)
{
}

void CPUSolver::connect(const uint8_t* img, const uint32_t height, const uint32_t width, int32_t* mappings)
{
  // 1. Initialize all active sites
  for (uint32_t i = 0; i < height; i++)
  {
    for (uint32_t j = 0; j < width; j++)
    {
      // Non-active sites are marked with -1
      if (img[i * width + j] == 0)
        mappings[i * width + j] = -1;
      else
        mappings[i * width + j] = i * width + j;
    }
  }

  // 2. Unite within windows and flatten (path compression)
  for (int32_t i = 0; i < (int32_t)height; i++){
    for (int32_t j = 0; j < (int32_t)width; j++){
      // Ignore inactive sites
      if (mappings[i * width + j] == -1)
        continue;

      // Get root of current pixel
      int32_t root = find(mappings, i, j, width);

      // Search within the window
      for (int32_t wi = i - m_vDist; wi <= i + (int32_t)m_vDist; wi++){
        // Ignore if out of range
        if (wi < 0 || wi >= (int32_t)height)
          continue;
        for (int32_t wj = j - m_hDist; wj <= j + (int32_t)m_hDist; wj++){
          // Ignore if out of range
          if (wj < 0 || wj >= (int32_t)width)
            continue;

          // Ignore if not active
          if (mappings[wi * width + wj] == -1)
            continue;

          // Otherwise, follow the path to root for candidate pixel
          int32_t candroot = find(mappings, wi, wj, width);

          // Unite if not already connected
          if (root != candroot){
            int32_t unitedroot = std::min(root, candroot);

            // Path compress the current pixel
            if (root != unitedroot)
              pathcompress(mappings, i, j, width, unitedroot);
            // Otherwise path compress the candidate pixel
            else
              pathcompress(mappings, wi, wj, width, unitedroot);
          }
        } // end window loops (columns)
      } // end window loops (rows)
    } // end pixel loops (columns)
  } // end pixel loops (rows)


}

void CPUSolver::connect(const std::vector<uint8_t>& img, const uint32_t height, const uint32_t width, std::vector<int32_t>& mappings){
  if (img.size() != (size_t)height * width)
    throw std::runtime_error("Invalid input image size");
  if (mappings.size() != img.size())
    throw std::runtime_error("Invalid mappings size; should equal input img size");

  this->connect(img.data(), height, width, mappings.data());
}

void CPUSolver::readout(const uint32_t height, const uint32_t width, std::vector<int32_t>& mapping){
  for (uint32_t i = 0; i < height; ++i){
    for (uint32_t j = 0; j < width; ++j){
      int32_t root = this->find(mapping.data(), i, j, width);
      mapping[i * width + j] = root;
    }
  }
}

int32_t CPUSolver::find(int32_t* mappings, const uint32_t i, const uint32_t j, const uint32_t width){
  if (mappings[i * width + j] < 0)
    return -1;

  int32_t root = mappings[i * width + j];
  while (root != mappings[root])
    root = mappings[root];

  return root;
}

void CPUSolver::pathcompress(int32_t* mappings, const uint32_t i, const uint32_t j, const uint32_t width, const int32_t root){
  int32_t* currentPtr = &mappings[i * width + j];
  int32_t currentParent = *currentPtr;
  while (currentParent != root){
    // Change the current pointer value
    *currentPtr = root;
    // Go to the parent
    currentPtr = &mappings[currentParent];
    // Update the parent
    currentParent = *currentPtr;
  }
}

}
